build.sh

```diff
@@ -22,16 +22,13 @@ ccache -p # print the config
 ccache -s # print the stats before reusing
 ccache -z # zero the stats
 
-go=$(which go)
-
 # Change Debug via  -DCMAKE_BUILD_TYPE=Debug
 cmake -DCMAKE_BUILD_TYPE=Release \
   -B$root/build \
   -H$root \
   -GNinja \
   -DCMAKE_C_COMPILER=$CC \
   -DCMAKE_CXX_COMPILER=$CXX \
-  -DCMAKE_GO_BINARY=$(go) \
```

Comment:
So I did this because redpanda has its own internal build system. Within that, we use a specific go version as well, and the idea was to use this option to pass the specific version of go to cmake in internal builds, and for open source, just use the system.

I tried using `find_program` but it was giving me issues. Maybe I can give these changes a shot

Reply:
The idea is to not have `find_program` choose the system installed version of `go` for our developer builds, and only falling back to the system version if the developer installed version of `go` is not found

Reply:
> a specific go version as well

could you share the path in which the specific go is located? so probably i can understand why the `/usr/bin/go` one wins?

Reply:
yeah we usually install `go` in `vbuild/go/<go_version>/bin` you can install `go` there and test out your branch

Reply:
I've tried passing this pass as a `HINT` to `find_program` but it always seemed to choose my system version in either case

Reply:
probably `HINT` is not the best way to customize the path. as CMake already provides a couple ways allowing user to specify additional paths in which to find the program.  i tested `$PATH`, `$CMAKE_PROGRAM_PATH` and `$CMAKE_PREFIX_PATH` env variables with my change using following steps:

```console
$ echo $PATH
/opt/rh/gcc-toolset-12/root/usr/bin:/home/kefu/.local/bin:/usr/share/Modules/bin:/usr/lib64/ccache:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
$ cd redpanda
$ mkdir -p vbuild/go/1.17/bin
$ ln -s /usr/bin/go vbuild/go/1.17/bin/go
$ old_path=$PATH
$ PATH=$PWD/vbuild/go/1.17/bin:$PATH
$ rm build/CMakeCache.txt
$ cmake -B$PWD/build -H$PWD -DCMAKE_BUILD_TYPE=Debug -GNinja \
  -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ \
  -DDEPOT_TOOLS_DIR=/opt/depot_tools -DCMAKE_EXE_LINKER_FLAGS=-fuse-ld=lld ..
$ grep CMAKE_GO_BINARY build/CMakeCache.txt
CMAKE_GO_BINARY:PATH=/home/kefu/dev/redpanda/vbuild/go/1.17/bin/go
$ rm build/CMakeCache.txt
```

use `CMAKE_PROGRAM_PATH` environment variables as proposed by CMake document
```console
$ PATH=$old_path
$ CMAKE_PROGRAM_PATH=$PWD/vbuild/go/1.17/bin \
  cmake -B$PWD/build -H$PWD -DCMAKE_BUILD_TYPE=Debug -GNinja \
  -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ \
  -DDEPOT_TOOLS_DIR=/opt/depot_tools -DCMAKE_EXE_LINKER_FLAGS=-fuse-ld=lld ..
$ grep CMAKE_GO_BINARY build/CMakeCache.txt
CMAKE_GO_BINARY:PATH=/home/kefu/dev/redpanda/vbuild/go/1.17/bin/go
```

or, use `CMAKE_PREFIX_PATH` environment variables as proposed by CMake document
```console
$ CMAKE_PREFIX_PATH=$PWD/vbuild/go/1.17 \
  cmake -B$PWD/build -H$PWD -DCMAKE_BUILD_TYPE=Debug -GNinja \
  -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ \
  -DDEPOT_TOOLS_DIR=/opt/depot_tools -DCMAKE_EXE_LINKER_FLAGS=-fuse-ld=lld ..
$ grep CMAKE_GO_BINARY build/CMakeCache.txt
CMAKE_GO_BINARY:PATH=/home/kefu/dev/redpanda/vbuild/go/1.17/bin/go
```

Reply:
Looks good to me, might as well also change the variable name to `GO_EXECUTABLE`. Let me know when you've prepared the changes and I will test against our internal build system, making the necessary changes to modify the path to the locally installed version of go

Reply:
sure. will get back to this tomorrow.

Reply:
@graphcareful hi Rob, i added a commit to change the variable name to `GO_EXECUTABLE` and tested the changes like:

```console
$ rm build/CMakeCache.txt
$ PATH=$PWD/vbuild/go/1.17/bin:$PATH cmake -B$PWD/build -H$PWD -DCMAKE_BUILD_TYPE=Debug -GNinja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DDEPOT_TOOLS_DIR=/opt/depot_tools -DCMAKE_EXE_LINKER_FLAGS=-fuse-ld=lld ..
$ grep GO_EXECUTABLE build/CMakeCache.txt
GO_EXECUTABLE:FILEPATH=/home/kefu/dev/redpanda/vbuild/go/1.17/bin/go
```

Reply:
Hmm its not working for me, maybe i've done something wrong? Here's how i've configured things:
```
task: [rp:configure] PATH=/home/robert/workspace/redpanda/vbuild/llvm/install/bin:$PATH                                                            
CMAKE_PROGRAM_PATH=$PATH:/home/robert/workspace/redpanda/vbuild/go/1.17.8/bin/go                                                                   
cmake \                                                                                                                                            
  -GNinja \                                                                                                                                        
  -DREDPANDA_DEPS_INSTALL_DIR="/home/robert/workspace/redpanda/vbuild/debug/clang/rp_deps_install" \                                               
  -DVECTORIZED_CMAKE_DIR="/home/robert/workspace/redpanda/vtools/cmake" \                                                                          
  -DDEPOT_TOOLS_DIR="/home/robert/workspace/redpanda/vbuild/depot_tools" \                                                                         
  -DCMAKE_BUILD_TYPE=Debug \                                                                                                                       
  -DCMAKE_C_COMPILER=clang \                                                                                                                       
  -DCMAKE_CXX_COMPILER=clang++ \                                                                                                                   
  -DCMAKE_PROGRAM_PATH="/home/robert/.local/bin:/home/robert/bin:/usr/lib64/ccache:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/var/lib/snapd
/snap/bin:/home/robert/.cabal/bin:/home/robert/bin:/home/robert/go/bin:/home/robert/workspace/redpanda/vbuild/llvm/install/bin:/home/robert/workspa
ce/redpanda/vbuild/go/linux/bin:/home/robert/workspace/redpanda/vbuild/go/1.17.8/bin/go" \                                                           -DCMAKE_LD_PATH= \                                                                                                                               
  -DV_DEPS_SKIP_BUILD=OFF \                                                                                                                          -DV_DEPS_ONLY=OFF \                                                                                                                              
  -DV_ENABLE_COV=OFF \                                                                                                                               -DREDPANDA_ENABLE_V8="FALSE" \                                                                                                                     -DENABLE_GIT_VERSION=ON \                                                                                                                          -DWITH_DISTCC=OFF \                                                                                                                                -B/home/robert/workspace/redpanda/vbuild/debug/clang \                                                                                           
  -H/home/robert/workspace/redpanda                                                                                                                                                                                                                                                                   
```
Then when I probe for the value:
```
➜  redpanda git:(cmake-finds-go) ✗ grep GO_EXECUTABLE vbuild/debug/clang/CMakeCache.txt
GO_EXECUTABLE:FILEPATH=/usr/bin/go
```